### PR TITLE
Prefer remote replicas for primary promotion during migration

### LINKED ISSUE #13135: [Remote Store] Prefer remote replicas for primary promotion during migration
### Is your feature request related to a problem? Please describe

* During remote store migration (`MIXED` compatibility mode and `REMOTE_STORE` migration direction), we aim to continuously move forward. 
* To ensure this, in case of a primary shard failover:
  * If the primary and one or more replica copies had already migrated to remote nodes, then the new primary must be chosen randomly from the migrated replicas
  * If no migrated replica is found, fall back upon the existing mechanism

### Describe the solution you'd like

Inside `unassignPrimaryAndPromoteActiveReplicaIfExists` in `RoutingNodes.java`:
* `metadata` can be used to identify migration conditions
* Using the `failedShard`'s id, we can look for migrated replica copies and implement the above preference

### Related component

Storage:Remote

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ltaragi**
### Description
* During remote store migration (`MIXED` compatibility mode and `REMOTE_STORE` migration direction), we aim to continuously move forward. 
* To ensure this, in case of a primary shard failover:
  * If the primary and one or more replica copies had already migrated to remote nodes, then the new primary must be chosen randomly from the migrated replicas
  * If no migrated replica is found, fall back upon the existing priority choosing
* This preference has been implemented as follows:
  * `metadata` is used to identify migration conditions
  * `activeReplicaOnRemoteNode` makes use of the `failedShard`'s id, to obtain migrated replica copies and randomly returns one 

### Related Issues
Resolves #13135

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java [hunk 1]
```diff
@@ -418,6 +422,29 @@ public ShardRouting activeReplicaWithOldestVersion(ShardId shardId) {
...
+                RoutingNode nd = node(shr.currentNodeId());
+                return (nd != null && nd.node().isRemoteStoreNode());
+            })
+            .collect(Collectors.toList());
+        ShardRouting replicaShard = null;
+        if (replicaShardsOnRemote.isEmpty() == false) {
+            Random rand = Randomness.get();
+            replicaShard = replicaShardsOnRemote.get(rand.nextInt(replicaShardsOnRemote.size()));
+        }
+        return replicaShard;
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java:445]: Do we need to pick active replica with highest checkpoint rather than choosing a random one

Comment (ltaragi) [reply server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java:445]: Checkpoint information is not available at `unassignPrimaryAndPromoteActiveReplicaIfExists`, so we are picking one active replica randomly.

File: server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java [hunk 2]
```diff
@@ -735,11 +762,17 @@ private void unassignPrimaryAndPromoteActiveReplicaIfExists(
         RoutingChangesObserver routingChangesObserver
     ) {
         assert failedShard.primary();
-        ShardRouting activeReplica;
-        if (metadata.isSegmentReplicationEnabled(failedShard.getIndexName())) {
-            activeReplica = activeReplicaWithOldestVersion(failedShard.shardId());
-        } else {
-            activeReplica = activeReplicaWithHighestVersion(failedShard.shardId());
+        ShardRouting activeReplica = null;
+        if (isMigratingToRemoteStore(new ClusterSettings(metadata.settings(), ClusterSettings.BUILT_IN_CLUSTER_SETTINGS))) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java:766]: lets not create a cluster settings object here and use `metadata` itself to determine. We can create a new helper function for it .

File: server/src/test/java/org/opensearch/cluster/routing/allocation/FailedShardsRoutingTests.java [hunk 3]
```diff
@@ -812,4 +822,134 @@ private void testReplicaIsPromoted(boolean isSegmentReplicationEnabled) {
             }
         }
     }
+
+    public void testPreferReplicaOnRemoteNodeForPrimaryPromotion() {
```
Comment (gbbafna) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/FailedShardsRoutingTests.java:826]: can we add IT as well ?

File: server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java [hunk 4]
```diff
@@ -418,6 +422,29 @@ public ShardRouting activeReplicaWithOldestVersion(ShardId shardId) {
...
+            .filter((shr) -> {
+                RoutingNode nd = node(shr.currentNodeId());
+                return (nd != null && nd.node().isRemoteStoreNode());
+            })
+            .collect(Collectors.toList());
+        ShardRouting replicaShard = null;
+        if (replicaShardsOnRemote.isEmpty() == false) {
+            Random rand = Randomness.get();
+            replicaShard = replicaShardsOnRemote.get(rand.nextInt(replicaShardsOnRemote.size()));
+        }
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java:444]: we can just return the first replica .

File: server/src/main/java/org/opensearch/node/remotestore/RemoteStoreNodeService.java [hunk 5]
```diff
@@ -238,4 +239,17 @@ public static boolean isMigratingToRemoteStore(ClusterSettings clusterSettings)
...
         return (isMixedMode && isRemoteStoreMigrationDirection);
     }
+
+    /**
+     * To check if the cluster is undergoing remote store migration using metadata settings
+     * @return
+     * <code>true</code> For <code>REMOTE_STORE</code> migration direction and <code>MIXED</code> compatibility mode,
+     * <code>false</code> otherwise
+     */
+    public static boolean isMigratingToRemoteStore(Settings settings) {
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/node/remotestore/RemoteStoreNodeService.java:249]: this doesn't look correct. `Settings` is very generic and can be any settings (cluster/index). Can we take `Metadata` as an input ?

File: server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java [hunk 6]
```diff
@@ -735,11 +751,17 @@ private void unassignPrimaryAndPromoteActiveReplicaIfExists(
         RoutingChangesObserver routingChangesObserver
     ) {
         assert failedShard.primary();
-        ShardRouting activeReplica;
-        if (metadata.isSegmentReplicationEnabled(failedShard.getIndexName())) {
-            activeReplica = activeReplicaWithOldestVersion(failedShard.shardId());
-        } else {
-            activeReplica = activeReplicaWithHighestVersion(failedShard.shardId());
+        ShardRouting activeReplica = null;
+        if (isMigratingToRemoteStore(metadata)) {
```
Comment (shourya035) [inline server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java:755]: Will this be okay with indexes that has been created after the migration starts? For example, when an index is created after the migration starts (on the remote enabled nodes with remote store index settings), we should ideally treat that as just another segrep shard copy and use the `activeReplicaWithOldestVersion` check to ensure segrep based functionalities are honored.

tl;dr, should we also add a `hasRemoteIndexSettings` check with this?

Comment (shourya035) [reply server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java:755]: Discussed this with @gbbafna . We can go ahead without this check for now.

**Review by shourya035 — APPROVED** [review]

**Review by gbbafna — APPROVED** [review]
